llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,393 @@
+/// from metadata information.
+///
+/// The algorithm is the following:
+/// 1) Make a list of all structure types from module @M. The list also
+///    contains DoubleTy since it is optional as well.
+/// 2) Make from list a type graph which consists of nodes corresponding to
+///    types and directed edges between nodes. An edge from type A to type B
+///    corresponds to the fact that A is contained within B.
+///    Examples: B is a pointer to A, B is a struct containing field of type A.
+/// 3) For every known type with aspects propagate it's aspects over graph.
+///    Every propagation is a separate run of BFS algorithm.
+///
+/// Time complexity: O((V + E) * T) where T is the number of input types
+/// containing aspects.
+void propagateAspectsToOtherTypesInModule(
+    const Module &M, TypeToAspectsMapTy &TypesWithAspects) {
+  std::unordered_set<const Type *> TypesToProcess;
+  Type *DoubleTy = Type::getDoubleTy(M.getContext());
+
+  // 6 is taken from sycl/include/CL/sycl/aspects.hpp
+  static constexpr int AspectFP64 = 6;
+  TypesWithAspects[DoubleTy].insert(AspectFP64);
+
+  TypesToProcess.insert(DoubleTy);
+  for (Type *T : M.getIdentifiedStructTypes())
+    TypesToProcess.insert(T);
+
+  TypesEdgesTy Edges;
+  for (const Type *T : TypesToProcess) {
+    for (const Type *TT : T->subtypes()) {
+      // If TT = %A*** then we want to get TT = %A
+      while (TT->isPointerTy())
+        TT = TT->getPointerElementType();
+
+      // We are not interested in some types. For example, IntTy.
+      if (!TypesToProcess.count(TT))
+        continue;
+
+      Edges[TT].push_back(T);
+    }
+  }
+
+  TypeToAspectsMapTy Result;
+  for (const Type *T : TypesToProcess) {
+    propagateAspectsThroughTypes(Edges, T, TypesWithAspects);
+  }
+}
+
+/// Returns all aspects which might be reached from type @T.
+/// It encompases composite structures and pointers.
+/// NB! This function inserts new records in @Types map for new discovered
+/// types. For the best perfomance pass this map in the next invocations.
+AspectsSetTy getAspectsFromType(const Type *T, TypeToAspectsMapTy &Types) {
+  auto It = Types.find(T);
+  if (It != Types.end())
+    return It->second;
+
+  // This is essential to no get into infinite recursive loops.
+  Types[T] = {};
+  AspectsSetTy Result;
+
+  for (const Type *TT : T->subtypes()) {
+    AspectsSetTy Aspects = getAspectsFromType(TT, Types);
+    Result.insert(Aspects.begin(), Aspects.end());
+  }
+
+  Types[T] = Result;
+  return Result;
+}
+
+/// Returns aspects which might be used in instruction @I.
+/// Function inspects return type and all operand's types.
+/// NB! This function inserts new records in @Types map for new discovered
+/// types. For the best perfomance pass this map in the next invocations.
+AspectsSetTy getAspectsUsedByInstruction(const Instruction &I,
+                                         TypeToAspectsMapTy &Types) {
+  Type *ReturnType = I.getType();
```

Comment:
What if this type is a pointer?

Reply:
Ok, sorry, I thought that `subtypes()` returns types for structs only. But I've looked into `llvm::Type`'s implementation and now I see that all types are covered.

Reply:
`composite_types1.ll` checks that since alloca returns a pointer.